### PR TITLE
Adding and registering an empty SW to enable PWA detection by Chrome

### LINKED ISSUE #55: add default Service Worker (for Chrome PWA detection)
can be blank for now (or contain a comment with a link to [Workbox](https://github.com/GoogleChrome/workbox)).

to make exported games PWA-able, at the minimum, a `sw.js` file must exist and the page must call `navigator.serviceWorker.register` to register it.

we can update it to add full-blown Service Worker offline capabilities later.

### PR DESCRIPTION
**Pull request by delapuente**
Fix #55

### CODE REVIEW

**Comment by delapuente** [comment]
It seems the default server that Unity set up is not serving the sw.js file correctly which is problematic. Anyway, a regular static server does it properly and the SW is registering.

File: Assets/WebGLTemplates/WebVR/webvr.js [hunk 1]
```diff
@@ -20,6 +20,8 @@
   var gamepads = [];
   var vrGamepads = [];
 
+  navigator.serviceWorker.register('./sw.js');
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/webvr.js]: can you add some logic for checking for the presence of `navigator.serviceWorker`?

if you want to flesh it out a bit more, I'd recommend something like this:

```js
  if ('serviceWorker' in navigator && window.isSecureContext === true) {
    if (navigator.serviceWorker.controller) {
      console.log('Running active Service Worker (controller: %s)', navigator.serviceWorker.controller.scriptURL);
    } else {
      navigator.serviceWorker.register('/sw.js').then(function (registration) {
        console.log('Successfully registered Service Worker (scope: %s)', registration.scope);
      }, function (err) {
        console.warn('Failed to register Service Worker:\n', err);
      });
    }
  }
```

File: Assets/WebGLTemplates/WebVR/sw.js [hunk 2]
```diff
@@ -0,0 +1 @@
+
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/sw.js]: can you add a short comment explaining why the Service Worker is empty (i.e., to trigger Chrome's heuristics for PWA)?

File: Assets/WebGLTemplates/WebVR/index.html [hunk 3]
```diff
@@ -3,8 +3,8 @@
   <head>
     <meta charset="utf-8">
     <meta name="viewport" content="width=device-width, initial-scale=1.0">
-    <title>%UNITY_WEB_NAME%</title>
-    <meta name="description" content="%UNITY_WEB_NAME%">
+    <title>%UNITY_CUSTOM_NAME% | %UNITY_WEB_NAME%</title>
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/index.html]: I'm running Unity `2017.2.0f3`. where are `%UNITY_CUSTOM_NAME%` and `%UNITY_CUSTOM_DESCRSIPTION%` defined?

I see there's support for `%UNITY_CUSTOM_…%` in the `Template Tags` section on [Unity's `Using WebGL Template` doc page](https://docs.unity3d.com/Manual/webgl-templates.html).

was this supposed to be added in a `.meta` file?

Comment (cvan) [reply Assets/WebGLTemplates/WebVR/index.html]: screen shot 2018-02-01 at 4 26 55 pm
screen shot 2018-02-01 at 4 29 39 pm

**Comment by cvan** [comment]
@delapuente:

> It seems the default server that Unity set up is not serving the sw.js file correctly which is problematic. Anyway, a regular static server does it properly and the SW is registering.

`/sw.js` seems to be served correctly for me when I run Unity's server from `Build Settings > Build and Run`:

```
% curl 'http://localhost:51549/sw.js' -I
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: 0
Server: Mono-HTTPAPI/1.0
Date: Fri, 02 Feb 2018 00:32:56 GMT
Content-Length: 1
```

when I run `http-server -c-1 -p9898`, I also see `/Build/sw.js` correctly load:

```
% curl http://localhost:9898/Build/sw.js -I
HTTP/1.1 200 OK
server: ecstatic-2.2.1
last-modified: Fri, 02 Feb 2018 00:26:03 GMT
etag: "8594164758-1-"2018-02-02T00:26:03.087Z""
cache-control: max-age=-1
content-length: 1
content-type: application/javascript; charset=UTF-8
Date: Fri, 02 Feb 2018 00:34:41 GMT
Connection: keep-alive
```

are you seeing something different?

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 1 above)
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/webvr.js]: also, we should check if the page is served from a secure origin (e.g., `localhost`, `127.x.x`, `https:`, etc.).

image

I'd recommend also checking [`window.isSecureContext === true`](https://www.w3.org/TR/powerful-features/#monkey-patching-global-object)

**Comment by cvan** [comment]
also, when pulling this branch and re-building the sample project in Unity, I see these untracked files in Git:

```
% gs                                /opt/unity-webvr-export (delapuente-add-sw)
On branch delapuente-add-sw
Your branch is up to date with 'delapuente/add-sw'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   Build/Build/Build.asm.code.unityweb
	modified:   Build/Build/Build.asm.framework.unityweb
	modified:   Build/Build/Build.asm.memory.unityweb
	modified:   Build/Build/Build.data.unityweb
	modified:   Build/Build/Build.json
	modified:   Build/Build/Build.wasm.code.unityweb
	modified:   Build/Build/Build.wasm.framework.unityweb
	modified:   Build/Build/UnityLoader.js
	modified:   Build/index.html
	modified:   Build/webvr.js
	modified:   ProjectSettings/ProjectVersion.txt

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	Build/sw.js

no changes added to commit (use "git add" and/or "git commit -a")
```

feel free to commit the changes on your end

File: Assets/WebGLTemplates/WebVR/index.html (hunk 3 above)
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/index.html]: According to the [Unity manual](https://docs.unity3d.com/Manual/webgl-templates.html), they appear under the template selection (try choosing a different one and then coming back to the WebVR one).

custom-meta-tags

**Comment by delapuente** [comment]
I will push the new Build in a separated commit or in a PR if you prefer.

**Comment by delapuente** [comment]
Regarding the server, what happens is that the Unity's server is not sending the proper `content-type` for the service worker so the browser can not register it.

I'm receiving this:

```
λ curl http://localhost:57582/sw.js -I
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: 0
Server: Mono-HTTPAPI/1.0
Date: Fri, 02 Feb 2018 08:22:21 GMT
Content-Length: 243
```
And getting this error in Chrome:
mime-type-error

**Comment by delapuente** [comment]
@cvan I've addressed the comments you left me. Can you review it again?

File: Assets/WebGLTemplates/WebVR/webvr.js [hunk 4]
```diff
@@ -20,6 +20,21 @@
   var gamepads = [];
   var vrGamepads = [];
 
+  if ('isSecureContext' in window && !window.isSecureContext) {
+    console.warn('The site is insecure: service workers will not work and the site will not be recognized as a PWA');
+  }
+  else if ('serviceWorker' in navigator) {
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/webvr.js]: feel free to move this up a line

File: Assets/WebGLTemplates/WebVR/webvr.js [hunk 5]
```diff
@@ -20,6 +20,21 @@
   var gamepads = [];
   var vrGamepads = [];
 
+  if ('isSecureContext' in window && !window.isSecureContext) {
+    console.warn('The site is insecure: service workers will not work and the site will not be recognized as a PWA');
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/webvr.js]: nits:
- `:` -> `;`
- `service workers` -> `Service Workers`

File: Assets/WebGLTemplates/WebVR/webvr.js [hunk 6]
```diff
@@ -20,6 +20,21 @@
   var gamepads = [];
   var vrGamepads = [];
 
+  if ('isSecureContext' in window && !window.isSecureContext) {
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/webvr.js]: personally, I'd not throw this warning if Service Workers aren't even supported (i.e., 'serviceWorker' in navigator`)

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 4 above)
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/webvr.js]: I'll do since the other branches are in the other style.

**Pull request merged**